Fix latency recording logic

dkropachev opened the pull request:
it is based on https://github.com/scylladb/scylla-bench/pull/51

 main.go and modes.go: fix latency registration logic
      Fix is needed due to the couple of issues, one of which is #50
      Another issue is that we do not destinct timeout error from other errors
      s-b did not report any latency if error occured, which is not correct in case of timeout errors.
      Since timeout errors can come from gocql and from the coordinator
      it was decided to distinct errors on timeout and non-timeout errors
      relaying on time it took for the workload to get an error
    
      Another change was not to interpolate latencies to avoid data spoiling, but store raw value instead

sitano reviewed (commented):
This 465a3529103999e5573b8401cd1059115e412791 patch do at least 2 separate things:

- Deletes my fix for Coordinated Omission
- Adds timeout

Those must be 2 separate patches.

Then I don't think that it is a good idea to delete interpolation. It only makes sense if you want results without Coordinated Omission (CO) being fixed. The only problem is that it will produce false latency results and would not generate correct throughput.

> Another change was not to interpolate latencies to avoid data spoiling,

I would suggest not removing CO related code but rather introduce a mode (maybe default) in which CO is disabled. However, everyone will probably run in incorrectly then. So it is also not a perfect solution IMHO.

I would even suggest adding a mode without CO, not removing CO, but keeping CO enabled mode by def. So that the users will have to explicitly state that they want incorrect results.

> but store raw value instead

Didn't look into that. But sounds weird.  Don't think there is a sense to store timeouts that tend to infinity.

sitano commented:
The code then may be split into 2 separate functionalities. 1 for generating throughput schedule and another one for latencies measurements where you can choose if you want to save interpolated values or not. Maybe even save both into 2 separate Histograms as it is done in YCSB or C-S (don't remember). But barely the correct requests schedule for throughput may be generated without `Expect()` content.

dkropachev commented:
After discussion it was decided to do following thig:
1. Limit latency that does go to histogram to timeout
2. Start registrying timeouts in latency histogram.

sitano commented:
> After discussion it was decided to do following thig:
> 
>     1. Limit latency that does go to histogram to timeout

If corrected or original latency > histogram maximum - we write maximum. 

>     2. Start registrying timeouts in latency histogram.

A small correction:

In place of (2) introduce 2 modes:

- in default mode latencies are registered as is
- if coordinated omission flag is enabled then we register latencies in 2 histograms - 1 for original metrics and 1 for corrected metrics. then we show both in the end. (and maybe in the process if requested)

dkropachev commented:
> > After discussion it was decided to do following thig:
> > ```
> > 1. Limit latency that does go to histogram to timeout
> > ```
> 
> If corrected or original latency > histogram maximum - we write maximum.
> 
> > ```
> > 2. Start registrying timeouts in latency histogram.
> > ```
> 
> A small correction:
> 
> In place of (2) introduce 2 modes:
> 
> * in default mode latencies are registered as is
> * if coordinated omission flag is enabled then we register latencies in 2 histograms - 1 for original metrics and 1 for corrected metrics. then we show both in the end. (and maybe in the process if requested)

Found another issue - test worker is not sending results to the channel when facing errors.

So I limit the fix to fix following issues:
1. Test worker is not sending results to the channel when facing errors
2. Limit latency that does go to histogram to timeout
3. Start to register timeouts in latency histogram

dkropachev commented:
rebased and tested. Works perfectly:

```
 ./scylla-bench -workload=uniform -mode=read -replication-factor=1 -partition-count=10 -clustering-row-count=10000000 -clustering-row-size=5120 -rows-per-request=10 -concurrency=7 -max-rate=32000 -duration=5760m -nodes 172.17.0.4
Configuration
Mode:                    read
Workload:                uniform
Timeout:                 5s
Consistency level:       quorum
Partition count:         10
Clustering rows:         10000000
Clustering row size:     Fixed(5120)
Rows per request:        10
Provide upper bound:     false
IN queries:              false
No lower bound:          false
Page size:               1000
Concurrency:             7
Connections:             4
Maximum rate:            32000 op/s
Client compression:      true

 time   ops/s  rows/s errors max    99.9th 99th   95th   90th   median mean   
   1s    6429       0      0 803ms  801ms  793ms  763ms  720ms  403ms  398ms  
   2s    6347       0      0 1.6s   1.6s   1.6s   1.6s   1.5s   1.2s   1.2s   
   3s    6196       0      0 2.4s   2.4s   2.4s   2.4s   2.3s   2s     2s     
   4s    6536       0      0 3.2s   3.2s   3.2s   3.2s   3.1s   2.8s   2.8s   
   5s    7185       0      0 4s     4s     4s     4s     3.9s   3.6s   3.6s   
   6s    7310       0      0 4.8s   4.8s   4.7s   4.7s   4.7s   4.4s   4.4s   
   7s    6668       0      0 5.6s   5.6s   5.5s   5.5s   5.5s   5.1s   5.1s   
   8s    6769       0      0 6.3s   6.3s   6.3s   6.3s   6.3s   5.9s   5.9s   
   9s    6402       0      0 7.1s   7.1s   7.1s   7.1s   7s     6.7s   6.7s   
  10s    7468       0      0 7.9s   7.9s   7.9s   7.9s   7.8s   7.5s   7.5s   
  11s    6694       0      0 8.7s   8.7s   8.7s   8.7s   8.6s   8.3s   8.3s   
  12s    6554       0      0 9.5s   9.5s   9.5s   9.5s   9.4s   9.1s   9.1s   
  13s    5921       0      0 10.3s  10.3s  10.3s  10.3s  10.2s  9.9s   9.9s   
  14s    6364       0      0 11.1s  11.1s  11.1s  11.1s  11s    10.7s  10.7s  
  15s    6350       0      0 11.9s  11.9s  11.9s  11.9s  11.8s  11.5s  11.5s  
  16s    6776       0      0 12.7s  12.7s  12.7s  12.7s  12.6s  12.3s  12.3s  
  17s    6679       0      0 13.5s  13.5s  13.5s  13.5s  13.4s  13.1s  13.1s  
  18s    6574       0      0 14.3s  14.3s  14.3s  14.3s  14.2s  13.9s  13.9s  
  19s    5950       0      0 15s    15s    15s    15s    15s    14.7s  14.7s  
  20s    6466       0      0 15s    15s    15s    15s    15s    15s    15s    
  21s    6334       0      0 15s    15s    15s    15s    15s    15s    15s    
  22s    6310       0      0 15s    15s    15s    15s    15s    15s    15s    ^C
interrupted

Results
Time (avg):      22.506230155s
Total ops:       147224
Total rows:      0
Operations/s:    6541.477573071696
Rows/s:          0
Latency:
  max:           15.007219711s 
  99.9th:        15.007219711s 
  99th:          15.007219711s 
  95th:          15.007219711s 
  90th:          15.007219711s 
  median:        8.657043455s 
  mean:          8.611283617s

```
